### ISSUE TITLE
EMD Loss Function

### ISSUE THREAD

**Issue by ahariri13** [pkomiske/EnergyFlow#17]
How can I use the end as a loss function ? 

I tried using the following function : 

```
def loss_function(r1,labels):
    
    check=energyflow.emd.emd(labels.detach().cpu().numpy(),r1.detach().cpu().numpy(),     R=1, norm=False, beta=0.1, measure='euclidean', coords='cartesian', return_flow=False, gdim=None, mask=False,  n_iter_max=100000, periodic_phi=False, phi_col=2, empty_policy='error')

    return check
```

but running loss.backward() in PyTorch returns the following error: 

 numpy.'float64' object has no attribute 'backward'.  

Any help is appreciated ! 
Thank you in advance.

**Comment by pkomiske** [comment]
Loss functions must be differentiable, at least in the standard backpropagation training paradigm that I assume PyTorch is using. You are only evaluating the EMD here, and PyTorch's automatic differentiation has no idea how to differentiate that, since the result is from a complicated algorithm. So what you are trying to do is not so simple (though it is a very nice idea!).

I would say that it's an open research question to demonstrate successful use of the Energy Mover's Distance as a loss function. Perhaps take a look at [this repository](https://github.com/Jackadsa/EMD-VAE), which uses an approximation of the EMD as a loss function for a variational autoencoder.
